src/core/Trail.tsx

```diff
@@ -0,0 +1,125 @@
+import { useFrame, useThree } from '@react-three/fiber'
+import * as React from 'react'
+import { ColorRepresentation, Group, Object3D, Vector2, Vector3 } from 'three'
+
+// @ts-ignore
+import { MeshLine, MeshLineMaterial } from 'meshline'
+
+interface TrailProps {
+  width?: number
+  color?: ColorRepresentation
+  length?: number
+  decay?: number
+  /**
+   * Wether to use the target's world or local positions
+   */
+  local?: boolean
+  attenuation?: (width: number) => number
+  target?: React.MutableRefObject<Object3D>
+}
+
+const shiftLeft = (collection, steps = 1) => {
+  collection.set(collection.subarray(steps))
+  collection.fill(-Infinity, -steps)
+  return collection
+}
+
+export function useTrail(target: Object3D, length: number = 1, decay: number = 1, local = false) {
```

Comment:
I don't like how the hook uses the object while the component version uses a ref, the Object + state is useful in the hook since I want to re-calc the initial positions on target change ( especially for the initial state).

Suggestions & changes welcome

Reply:
There's actually little reason if any for the component to use a ref as the `target`. We can just as easily use non-ref here:

https://github.com/FarazzShaikh/drei/blob/b1d779d0df49d42e4cedc72067c740d9541f549a/src/core/Trail.tsx#L73

Reply:
We could make it both too

```
target = Ref | Object3d

and in checks

target?.current || target 
```

Reply:
But do feel free to change it back to ref, as long as it works it's no biggie

Reply:
Yep that works too, lets wait on what the other reviewers have to say